streams/src/main/java/org/apache/kafka/streams/state/PrefixScanKeyValueStore.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state;
+
+/**
+ * A key-value store that only supports read operations.
+ * Implementations should be thread-safe as concurrent reads and writes are expected.
+ *
+ * Please note that this contract defines the thread-safe read functionality only; it does not
+ * guarantee anything about whether the actual instance is writable by another thread, or
+ * whether it uses some locking mechanism under the hood. For this reason, making dependencies
+ * between the read and write operations on different StateStore instances can cause concurrency
+ * problems like deadlock.
+ *
+ * @param <K> the key type
+ * @param <V> the value type
+ */
+public interface PrefixScanKeyValueStore<K, V> {
+    /**
+     * Return an iterator over all keys that share the same prefix byte[].
+     *
+     * @param prefix The byte[] prefix to perform the prefixScan for
```

Comment:
note this isn't a byte[], but a K. Not actually sure if this API makes sense.
For example, what would this API mean if my key type is UUID or MyCustomRecord?
In other words, only some key types even have prefixes.

Reply:
Corrected the comments - this is probably also related to your question about where should the prefixScan even go.
The only real requirements are RocksDB, Logging and Caching stores. Since they're so intertwined, it's hard for me to give you a good answer on if this should be here or not. If we have to ask, probably not, but TBH I'm really looking for guidance on this since I'm a bit over my head on where it should go. I have asked for feedback on this previously, but so far you're the only one to take not.